File: modulo2/shopapp/app/src/main/java/com/shopapp/presentation/ui/admin/usuarios/UsersAdminScreen.kt
// presentation/ui/admin/users/UsersAdminScreen.kt
package com.shopapp.presentation.ui.admin.users

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.LazyRow
import androidx.compose.foundation.lazy.items
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.*
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.hilt.navigation.compose.hiltViewModel
import com.shopapp.domain.model.User
import com.shopapp.presentation.components.LoadingScreen
import com.shopapp.presentation.components.ErrorScreen
import com.shopapp.presentation.viewmodel.UserRoleFilter
import com.shopapp.presentation.viewmodel.UsersAdminViewModel
import com.shopapp.theme.*

@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun UsersAdminScreen(
    viewModel: UsersAdminViewModel = hiltViewModel(),
) {
    val state     by viewModel.state.collectAsState()
    val filtered  by viewModel.filtered.collectAsState()
    val formState by viewModel.formState.collectAsState()

    var showForm    by remember { mutableStateOf(false) }
    var editTarget  by remember { mutableStateOf<User?>(null) }
    var deleteTarget by remember { mutableStateOf<User?>(null) }

    Column(
        modifier = Modifier
            .fillMaxSize()
            .background(Background),
    ) {
        // ── Header ────────────────────────────────────────────
        Surface(color = Surface, tonalElevation = 0.dp) {
            Column(modifier = Modifier.padding(16.dp)) {
                Row(
                    modifier              = Modifier.fillMaxWidth(),
                    horizontalArrangement = Arrangement.SpaceBetween,
                    verticalAlignment     = Alignment.CenterVertically,
                ) {
                    Column {
                        Text(
                            "Usuarios",
                            style      = MaterialTheme.typography.headlineMedium,
                            fontWeight = FontWeight.Bold,
                            color      = TextPrimary,
                        )
                        Text(
                            "${state.total} usuarios",
                            style = MaterialTheme.typography.bodySmall,
                            color = TextSecondary,
                        )
                    }
                    Row(horizontalArrangement = Arrangement.spacedBy(8.dp)) {
                        IconButton(onClick = viewModel::load) {
                            Icon(Icons.Default.Refresh, null, tint = TextSecondary)
                        }
                        Button(
                            onClick = { editTarget = null; showForm = true },
                            colors  = ButtonDefaults.buttonColors(
                                containerColor = Accent, contentColor = AccentOnDark,
                            ),
                            shape          = MaterialTheme.shapes.medium,
                            contentPadding = PaddingValues(horizontal = 16.dp, vertical = 10.dp),
                        ) {
                            Icon(Icons.Default.PersonAdd, null, modifier = Modifier.size(18.dp))
                            Spacer(Modifier.width(4.dp))
                            Text("Nuevo", fontWeight = FontWeight.Bold)
                        }
                    }
                }

                Spacer(Modifier.height(12.dp))

                // Búsqueda
                OutlinedTextField(
                    value         = state.search,
                    onValueChange = viewModel::setSearch,
                    placeholder   = { Text("Buscar usuario o email...", color = TextFaint) },
                    leadingIcon   = { Icon(Icons.Default.Search, null, tint = TextSecondary) },
                    singleLine    = true,
                    modifier      = Modifier.fillMaxWidth(),
                    shape         = MaterialTheme.shapes.medium,
                    colors        = OutlinedTextFieldDefaults.colors(
                        focusedBorderColor   = Accent,
                        unfocusedBorderColor = Border,
                        cursorColor          = Accent,
                    ),
                )

                Spacer(Modifier.height(10.dp))

                // Filtros de rol
                LazyRow(horizontalArrangement = Arrangement.spacedBy(8.dp)) {
                    items(UserRoleFilter.entries) { filter ->
                        FilterChip(
                            selected = state.roleFilter == filter,
                            onClick  = { viewModel.setRoleFilter(filter) },
                            label    = { Text(filter.label, style = MaterialTheme.typography.labelSmall) },
                            colors   = FilterChipDefaults.filterChipColors(
                                selectedContainerColor = Accent,
                                selectedLabelColor     = AccentOnDark,
                                containerColor         = Surface2,
                                labelColor             = TextSecondary,
                            ),
                        )
                    }
                }
            }
        }

        // ── Contenido ─────────────────────────────────────────
        when {
            state.isLoading -> LoadingScreen("Cargando usuarios...")
            state.error != null -> ErrorScreen(state.error!!, onRetry = viewModel::load)
            filtered.isEmpty() -> {
                Box(Modifier.fillMaxSize(), Alignment.Center) {
                    Column(horizontalAlignment = Alignment.CenterHorizontally) {
                        Text("👤", fontSize = 48.sp)
                        Spacer(Modifier.height(12.dp))
                        Text(
                            if (state.search.isBlank() && state.roleFilter == UserRoleFilter.ALL)
                                "Sin usuarios" else "Sin resultados",
                            style      = MaterialTheme.typography.titleMedium,
                            fontWeight = FontWeight.Bold,
                            color      = TextPrimary,
                        )
                    }
                }
            }
            else -> {
                LazyColumn(
                    modifier       = Modifier.fillMaxSize(),
                    contentPadding = PaddingValues(16.dp),
                    verticalArrangement = Arrangement.spacedBy(10.dp),
                ) {
                    items(filtered, key = { it.id }) { user ->
                        UserAdminCard(
                            user         = user,
                            onToggleStaff  = { viewModel.toggleStaff(user.id, !user.isStaff) },
                            onToggleActive = { viewModel.toggleActive(user.id) },
                            onEdit         = { editTarget = user; showForm = true },
                            onDelete       = { deleteTarget = user },
                        )
                    }
                }
            }
        }
    }

    // Bottom Sheet formulario
    if (showForm) {
        UserFormSheet(
            initial   = editTarget,
            formState = formState,
            onSave    = { payload ->
                if (editTarget != null) viewModel.updateUser(editTarget!!.id, payload)
                else viewModel.createUser(payload)
            },
            onDismiss = {
                showForm   = false
                editTarget = null
                viewModel.resetFormState()
            },
        )
    }

    // Diálogo de eliminación
    deleteTarget?.let { user ->
        AlertDialog(
            onDismissRequest = { deleteTarget = null },
            containerColor   = Surface,
            shape            = MaterialTheme.shapes.large,
            title            = { Text("¿Eliminar usuario?", color = TextPrimary) },
            text             = {
                Text(
                    "\"${user.username}\" se eliminará permanentemente. Esta acción no se puede deshacer.",
                    color = TextSecondary,
                )
            },
            confirmButton    = {
                TextButton(onClick = {
                    viewModel.deleteUser(user.id)
                    deleteTarget = null
                }) { Text("Eliminar", color = Error, fontWeight = FontWeight.Bold) }
            },
            dismissButton    = {
                TextButton(onClick = { deleteTarget = null }) {
                    Text("Cancelar", color = TextSecondary)
                }
            },
        )
    }
}

// ── UserAdminCard ─────────────────────────────────────────────

@Composable
private fun UserAdminCard(
    user:           User,
    onToggleStaff:  () -> Unit,
    onToggleActive: () -> Unit,
    onEdit:         () -> Unit,
    onDelete:       () -> Unit,
) {
    Surface(
        shape  = MaterialTheme.shapes.large,
        color  = if (user.isActive) Surface else Surface.copy(alpha = 0.55f),
        modifier = Modifier.fillMaxWidth(),
    ) {
        Row(
            modifier          = Modifier.padding(12.dp),
            verticalAlignment = Alignment.CenterVertically,
        ) {
            // Avatar
            Box(
                modifier         = Modifier
                    .size(46.dp)
                    .background(
                        brush = if (user.isStaff)
                            androidx.compose.ui.graphics.Brush.linearGradient(
                                listOf(Accent, AccentLight)
                            )
                        else
                            androidx.compose.ui.graphics.Brush.linearGradient(
                                listOf(Surface2, Border)
                            ),
                        shape = CircleShape,
                    ),
                contentAlignment = Alignment.Center,
            ) {
                Text(
                    text       = user.username.firstOrNull()?.uppercaseChar()?.toString() ?: "?",
                    color      = if (user.isStaff) AccentOnDark else TextSecondary,
                    fontWeight = FontWeight.Bold,
                    fontSize   = 18.sp,
                )
            }

            Spacer(Modifier.width(12.dp))

            // Info
            Column(modifier = Modifier.weight(1f)) {
                Row(
                    verticalAlignment     = Alignment.CenterVertically,
                    horizontalArrangement = Arrangement.spacedBy(6.dp),
                ) {
                    Text(
                        text       = user.username,
                        style      = MaterialTheme.typography.bodyLarge,
                        fontWeight = FontWeight.SemiBold,
                        color      = TextPrimary,
                    )
                    // Badge staff
                    if (user.isStaff) {
                        Surface(
                            color  = Accent.copy(alpha = 0.15f),
                            shape  = MaterialTheme.shapes.extraSmall,
                        ) {
                            Text(
                                "Staff",
                                color      = Accent,
                                fontSize   = 10.sp,
                                fontWeight = FontWeight.Bold,
                                modifier   = Modifier.padding(horizontal = 6.dp, vertical = 2.dp),
                            )
                        }
                    }
                    // Badge inactivo
                    if (!user.isActive) {
                        Surface(
                            color  = Error.copy(alpha = 0.12f),
                            shape  = MaterialTheme.shapes.extraSmall,
                        ) {
                            Text(
                                "Inactivo",
                                color      = Error,
                                fontSize   = 10.sp,
                                fontWeight = FontWeight.Bold,
                                modifier   = Modifier.padding(horizontal = 6.dp, vertical = 2.dp),
                            )
                        }
                    }
                }
                Text(
                    text  = user.email,
                    style = MaterialTheme.typography.bodySmall,
                    color = TextSecondary,
                )
                Text(
                    text  = "${user.numOrders} pedido${if (user.numOrders != 1) "s" else ""}",
                    style = MaterialTheme.typography.bodySmall,
                    color = Accent,
                    fontWeight = FontWeight.SemiBold,
                )
            }

            // Acciones
            Column(
                horizontalAlignment = Alignment.End,
                verticalArrangement = Arrangement.spacedBy(4.dp),
            ) {
                Row(horizontalArrangement = Arrangement.spacedBy(2.dp)) {
                    // Toggle staff
                    IconButton(onClick = onToggleStaff, modifier = Modifier.size(32.dp)) {
                        Icon(
                            imageVector = if (user.isStaff) Icons.Default.AdminPanelSettings
                            else Icons.Default.PersonOutline,
                            contentDescription = if (user.isStaff) "Quitar staff" else "Dar staff",
                            tint     = if (user.isStaff) Accent else TextFaint,
                            modifier = Modifier.size(18.dp),
                        )
                    }
                    // Toggle activo
                    IconButton(onClick = onToggleActive, modifier = Modifier.size(32.dp)) {
                        Icon(
                            imageVector = if (user.isActive) Icons.Default.ToggleOn
                            else Icons.Default.ToggleOff,
                            contentDescription = if (user.isActive) "Desactivar" else "Activar",
                            tint     = if (user.isActive) Success else TextFaint,
                            modifier = Modifier.size(22.dp),
                        )
                    }
                    // Editar
                    IconButton(onClick = onEdit, modifier = Modifier.size(32.dp)) {
                        Icon(
                            Icons.Default.Edit, contentDescription = "Editar",
                            tint = TextSecondary, modifier = Modifier.size(18.dp),
                        )
                    }
                    // Eliminar
                    IconButton(onClick = onDelete, modifier = Modifier.size(32.dp)) {
                        Icon(
                            Icons.Default.PersonRemove, contentDescription = "Eliminar",
                            tint = Error, modifier = Modifier.size(18.dp),
                        )
                    }
                }
            }
        }
    }
}

File: modulo4/flutter_shop_app/lib/presentation/screens/admin/users_admin_screen.dart
// lib/presentation/screens/admin/users_admin_screen.dart

import 'package:flutter/material.dart';
import 'package:flutter_riverpod/flutter_riverpod.dart';
import '../../../theme/app_colors.dart';
import '../../../domain/model/user.dart';
import '../../providers/users_admin_provider.dart';
import '../../widgets/user_form.dart';

class UsersAdminScreen extends ConsumerWidget {
  const UsersAdminScreen({super.key});

  @override
  Widget build(BuildContext context, WidgetRef ref) {
    final state    = ref.watch(usersAdminProvider);
    final filtered = state.filtered;
    final tt       = Theme.of(context).textTheme;

    return Column(
      children: [
        // ── Header ──────────────────────────────────────────
        Container(
          color:   AppColors.surface,
          padding: const EdgeInsets.fromLTRB(16, 16, 16, 0),
          child:   Column(
            children: [
              Row(
                mainAxisAlignment: MainAxisAlignment.spaceBetween,
                children: [
                  Column(
                    crossAxisAlignment: CrossAxisAlignment.start,
                    children: [
                      Text('Usuarios', style: tt.headlineMedium),
                      Text('${state.total} usuarios',
                          style: const TextStyle(color: AppColors.textSecondary, fontSize: 13)),
                    ],
                  ),
                  Row(
                    children: [
                      IconButton(
                        onPressed: () => ref.read(usersAdminProvider.notifier).load(),
                        icon: const Icon(Icons.refresh_rounded, color: AppColors.textSecondary),
                      ),
                      ElevatedButton.icon(
                        onPressed: () => showUserForm(context, ref),
                        icon:      const Icon(Icons.person_add_outlined, size: 18),
                        label:     const Text('Nuevo'),
                        style:     ElevatedButton.styleFrom(
                          minimumSize: const Size(0, 40),
                          padding:     const EdgeInsets.symmetric(horizontal: 16),
                        ),
                      ),
                    ],
                  ),
                ],
              ),
              const SizedBox(height: 12),

              // Búsqueda
              TextField(
                onChanged:  ref.read(usersAdminProvider.notifier).setSearch,
                decoration: const InputDecoration(
                  hintText:   'Buscar usuario o email...',
                  prefixIcon: Icon(Icons.search_rounded, color: AppColors.textSecondary),
                  contentPadding: EdgeInsets.symmetric(vertical: 10),
                ),
                style: const TextStyle(color: AppColors.textPrimary),
              ),
              const SizedBox(height: 10),

              // Chips de filtro de rol
              SizedBox(
                height: 34,
                child:  ListView(
                  scrollDirection: Axis.horizontal,
                  children: UserRoleFilter.values.map((f) => Padding(
                    padding: const EdgeInsets.only(right: 8),
                    child:   ChoiceChip(
                      label:     Text(f.label),
                      selected:  state.roleFilter == f,
                      onSelected:(_) =>
                          ref.read(usersAdminProvider.notifier).setRoleFilter(f),
                    ),
                  )).toList(),
                ),
              ),
              const SizedBox(height: 12),
            ],
          ),
        ),

        // ── Lista ─────────────────────────────────────────────
        Expanded(
          child: Builder(builder: (_) {
            if (state.isLoading) {
              return const Center(
                child: CircularProgressIndicator(color: AppColors.accent),
              );
            }
            if (state.error != null) {
              return Center(
                child: Column(
                  mainAxisAlignment: MainAxisAlignment.center,
                  children: [
                    Text(state.error!, style: const TextStyle(color: AppColors.error)),
                    const SizedBox(height: 12),
                    ElevatedButton(
                      onPressed: () => ref.read(usersAdminProvider.notifier).load(),
                      child: const Text('Reintentar'),
                    ),
                  ],
                ),
              );
            }
            if (filtered.isEmpty) {
              return const Center(
                child: Column(
                  mainAxisAlignment: MainAxisAlignment.center,
                  children: [
                    Text('👤', style: TextStyle(fontSize: 48)),
                    SizedBox(height: 12),
                    Text('Sin usuarios',
                        style: TextStyle(
                          color: AppColors.textPrimary, fontSize: 18, fontWeight: FontWeight.bold,
                        )),
                  ],
                ),
              );
            }

            return ListView.separated(
              padding:         const EdgeInsets.all(16),
              itemCount:       filtered.length,
              separatorBuilder:(_, __) => const SizedBox(height: 10),
              itemBuilder: (_, i) => _UserCard(
                user:          filtered[i],
                onToggleStaff: () => ref.read(usersAdminProvider.notifier)
                    .toggleStaff(filtered[i].id, !filtered[i].isStaff),
                onToggleActive:() => ref.read(usersAdminProvider.notifier)
                    .toggleActive(filtered[i].id),
                onEdit:        () => showUserForm(context, ref, initial: filtered[i]),
                onDelete:      () => _confirmDelete(context, ref, filtered[i]),
              ),
            );
          }),
        ),
      ],
    );
  }

  void _confirmDelete(BuildContext context, WidgetRef ref, User user) {
    showDialog(
      context: context,
      builder: (_) => AlertDialog(
        backgroundColor: AppColors.surface,
        shape:           RoundedRectangleBorder(borderRadius: BorderRadius.circular(16)),
        title:           const Text('¿Eliminar usuario?',
            style: TextStyle(color: AppColors.textPrimary)),
        content:         Text(
          '"${user.username}" se eliminará permanentemente.',
          style: const TextStyle(color: AppColors.textSecondary),
        ),
        actions: [
          TextButton(
            onPressed: () => Navigator.pop(context),
            child:     const Text('Cancelar'),
          ),
          TextButton(
            onPressed: () {
              Navigator.pop(context);
              ref.read(usersAdminProvider.notifier).deleteUser(user.id);
            },
            child: const Text('Eliminar',
                style: TextStyle(color: AppColors.error, fontWeight: FontWeight.bold)),
          ),
        ],
      ),
    );
  }
}

// ── UserCard ──────────────────────────────────────────────────

class _UserCard extends StatelessWidget {
  final User         user;
  final VoidCallback onToggleStaff;
  final VoidCallback onToggleActive;
  final VoidCallback onEdit;
  final VoidCallback onDelete;

  const _UserCard({
    required this.user,
    required this.onToggleStaff,
    required this.onToggleActive,
    required this.onEdit,
    required this.onDelete,
  });

  @override
  Widget build(BuildContext context) => Opacity(
    opacity: user.isActive ? 1.0 : 0.55,
    child:   Container(
      padding:    const EdgeInsets.all(12),
      decoration: BoxDecoration(
        color:        AppColors.surface,
        borderRadius: BorderRadius.circular(14),
        border:       Border.all(color: AppColors.border),
      ),
      child: Row(
        children: [
          // Avatar
          Container(
            width:  46, height: 46,
            decoration: BoxDecoration(
              gradient: user.isStaff
                  ? const LinearGradient(colors: [AppColors.accent, AppColors.accentLight])
                  : LinearGradient(colors: [AppColors.surface2, AppColors.border]),
              shape: BoxShape.circle,
            ),
            child: Center(
              child: Text(
                user.username.isNotEmpty ? user.username[0].toUpperCase() : '?',
                style: TextStyle(
                  color:      user.isStaff ? AppColors.onAccent : AppColors.textSecondary,
                  fontWeight: FontWeight.bold, fontSize: 18,
                ),
              ),
            ),
          ),
          const SizedBox(width: 12),

          // Info
          Expanded(
            child: Column(
              crossAxisAlignment: CrossAxisAlignment.start,
              children: [
                Row(
                  children: [
                    Flexible(
                      child: Text(
                        user.username,
                        style: const TextStyle(
                          color: AppColors.textPrimary, fontWeight: FontWeight.w600,
                        ),
                        overflow: TextOverflow.ellipsis,
                      ),
                    ),
                    const SizedBox(width: 6),
                    if (user.isStaff)
                      Container(
                        padding: const EdgeInsets.symmetric(horizontal: 6, vertical: 1),
                        decoration: BoxDecoration(
                          color:        AppColors.accent.withValues(alpha: 0.15),
                          borderRadius: BorderRadius.circular(6),
                        ),
                        child: const Text('Staff',
                            style: TextStyle(
                              color: AppColors.accent, fontSize: 10, fontWeight: FontWeight.bold,
                            )),
                      ),
                    if (!user.isActive) ...[
                      const SizedBox(width: 4),
                      Container(
                        padding: const EdgeInsets.symmetric(horizontal: 6, vertical: 1),
                        decoration: BoxDecoration(
                          color:        AppColors.error.withValues(alpha: 0.12),
                          borderRadius: BorderRadius.circular(6),
                        ),
                        child: const Text('Inactivo',
                            style: TextStyle(
                              color: AppColors.error, fontSize: 10, fontWeight: FontWeight.bold,
                            )),
                      ),
                    ],
                  ],
                ),
                Text(user.email,
                    style: const TextStyle(color: AppColors.textSecondary, fontSize: 12),
                    overflow: TextOverflow.ellipsis),
                Text(
                  '${user.numOrders} pedido${user.numOrders != 1 ? "s" : ""}',
                  style: const TextStyle(
                    color: AppColors.accent, fontSize: 11, fontWeight: FontWeight.w600,
                  ),
                ),
              ],
            ),
          ),

          // Acciones
          Column(
            crossAxisAlignment: CrossAxisAlignment.end,
            children: [
              Row(
                mainAxisSize: MainAxisSize.min,
                children: [
                  // Toggle staff
                  GestureDetector(
                    onTap: onToggleStaff,
                    child: Padding(
                      padding: const EdgeInsets.all(4),
                      child:   Icon(
                        user.isStaff
                            ? Icons.admin_panel_settings
                            : Icons.person_outline,
                        color:  user.isStaff ? AppColors.accent : AppColors.textFaint,
                        size:   20,
                      ),
                    ),
                  ),
                  // Toggle activo
                  GestureDetector(
                    onTap: onToggleActive,
                    child: Padding(
                      padding: const EdgeInsets.all(4),
                      child:   Icon(
                        user.isActive ? Icons.toggle_on : Icons.toggle_off,
                        color:  user.isActive ? AppColors.success : AppColors.textFaint,
                        size:   26,
                      ),
                    ),
                  ),
                  // Editar
                  GestureDetector(
                    onTap: onEdit,
                    child: const Padding(
                      padding: EdgeInsets.all(4),
                      child:   Icon(Icons.edit_outlined,
                          color: AppColors.textSecondary, size: 20),
                    ),
                  ),
                  // Eliminar
                  GestureDetector(
                    onTap: onDelete,
                    child: const Padding(
                      padding: EdgeInsets.all(4),
                      child:   Icon(Icons.person_remove_outlined,
                          color: AppColors.error, size: 20),
                    ),
                  ),
                ],
              ),
            ],
          ),
        ],
      ),
    ),
  );
}